File: app/components/folio/console/links/modal/url_picker_component.rb
# frozen_string_literal: true

class Folio::Console::Links::Modal::UrlPickerComponent < Folio::Console::ApplicationComponent
  include Folio::Console::Cell::IndexFilters

  def initialize(url_json:)
    @url_json = url_json
  end

  def before_render
    if @record.nil? && @url_json[:record_id].present? && @url_json[:record_type].present?
      klass = @url_json[:record_type].safe_constantize

      if klass < ActiveRecord::Base
        record = klass.find_by(id: @url_json[:record_id])

        if can_now?(:show, record)
          @record = record
        end
      end
    end
  end

  def tabs
    @tabs ||= begin
      first_active = @record.present? || @url_json[:href].blank?

      [
        { label: t(".tab/pick"), key: :pick, active: first_active },
        { label: t(".tab/custom_url"), key: :custom_url, active: !first_active }
      ]
    end
  end

  def data
    stimulus_controller("f-c-links-modal-url-picker",
                        action: {
                          "f-c-links-modal-list:selectedRecord" => "selectedRecord",
                          "f-c-input-form-group-url:edit" => "edit",
                          "f-c-input-form-group-url:remove" => "remove",
                        },
                        values: {
                          value_loading: false,
                          list_loading: false,
                          value_present: !!@record,
                          api_value_url: controller.value_console_api_links_path,
                          api_list_url: controller.list_console_api_links_path,
                          filtering: false,
                          autofocus_input: tabs[1][:active],
                        })
  end

  def cancel_button_model
    {
      data: stimulus_data(action: { click: "cancelFilters" }, target: "cancelButton"),
      variant: :danger,
      icon: :close,
      class: "f-c-links-modal-url-picker__list-filters-cancel-button"
    }
  end

  def filter_form(&block)
    opts = {
      url: "#filter",
      method: :get,
      html: {
        class: "f-c-links-modal-url-picker__list-filters-form",
        data: stimulus_data(action: {
          change: "onFormChange",
          folio_select2_change: "onFormChange",
          submit: "onFormSubmit"
        }, target: "form"),
      },
    }

    simple_form_for("", opts, &block)
  end

  def class_names_collection
    (%w[Folio::Page] + Rails.application.config.folio_console_links_mapping.keys).uniq.filter_map do |class_name|
      klass = class_name.safe_constantize

      if klass && klass < ActiveRecord::Base && can_now?(:show, klass)
        [klass.model_name.human, klass.to_s]
      end
    end
  end

  def additional_filters_hash
    Rails.application.config.folio_console_links_additional_filters
  end

  def additional_select(f, key, data)
    url = controller.folio.select2_console_api_autocomplete_path(klass: data[:klass],
                                                                 scope: data[:scope],
                                                                 order_scope: data[:order_scope],
                                                                 slug: data[:slug],
                                                                 label_method: data[:label_method])

    select2_select(f, key, data, url:)
  end

  def label_for_key(key)
    t(".placeholder/#{key}", default: key.to_s)
  end
end

File: app/components/folio/console/links/modal/url_picker_component.js
window.Folio.Stimulus.register('f-c-links-modal-url-picker', class extends window.Stimulus.Controller {
  static values = {
    valueLoading: Boolean,
    listLoading: Boolean,
    valuePresent: Boolean,
    filtering: Boolean,
    autofocusInput: Boolean,
    apiValueUrl: String,
    apiListUrl: String,
    absoluteUrls: { type: Boolean, default: false }
  }

  static targets = ['valueWrap', 'input', 'cancelButton', 'form', 'listContent']

  connect () {
    if (this.autofocusInputValue) {
      this.inputTarget.focus()

      const length = this.inputTarget.value.length
      this.inputTarget.setSelectionRange(length, length)
    }

    this.onQueryInput = Folio.debounce(() => {
      this.loadList()
    })
  }

  disconnect () {
    if (this.listLoadTimeout) {
      window.clearTimeout(this.listLoadTimeout)
      this.listLoadTimeout = null
    }

    if (this.listAbortController) {
      this.listAbortController.abort()
      delete this.listAbortController
    }

    if (this.valueLoadTimeout) {
      window.clearTimeout(this.valueLoadTimeout)
      this.valueLoadTimeout = null
    }

    if (this.valueAbortController) {
      this.valueAbortController.abort()
      delete this.valueAbortController
    }

    delete this.onQueryInput
  }

  selectedRecord (e) {
    this.inputTarget.value = e.detail.urlJson.href
    this.loadValue(e.detail.urlJson)
    this.dispatch('changed', { detail: { urlJson: e.detail.urlJson } })
  }

  loadValue (urlJson) {
    this.valueLoadingValue = true
    this.valueWrapTarget.innerHTML = ''

    if (this.valueAbortController) {
      this.valueAbortController.abort()
    }

    this.valueAbortController = new AbortController()

    const url = window.Folio.addParamsToUrl(this.apiValueUrlValue, { url_json: JSON.stringify(urlJson) })

    window.Folio.Api.apiGet(url, null, this.valueAbortController.signal).then((res) => {
      this.valueWrapTarget.innerHTML = res.data || ''
      this.valuePresentValue = !!res.data
      this.valueLoadingValue = false
    }).catch((e) => {
      this.valueLoadTimeout = window.setTimeout(() => {
        this.loadValue(urlJson)
      }, 1000)
    }).finally(() => {
      delete this.valueAbortController
    })
  }

  edit () {
    this.valuePresentValue = false
  }

  remove () {
    this.inputTarget.value = ""
    this.valuePresentValue = false
    this.dispatch('changed', { detail: { urlJson: { href: '', label: '' } } })
  }

  removeValue () {
    this.valuePresentValue = false
    this.valueWrapTarget.innerHTML = ''
  }

  onInputChange (e) {
    this.removeValue()

    const urlJson = { href: e.target.value }
    this.dispatch('changed', { detail: { urlJson } })
  }

  onFormChange (e) {
    this.markSelectsAndLoadList()
  }

  onFormSubmit (e) {
    e.preventDefault()
    this.markSelectsAndLoadList()
  }

  loadList () {
    const rawData = window.Folio.formToHash(this.formTarget)

    if (rawData.q && rawData.q.length < 3) {
      delete rawData.q
    }

    Object.keys(rawData).forEach((key) => {
      if (rawData[key] === '') {
        delete rawData[key]
      }
    })

    const json = JSON.stringify(rawData)

    if (json === "{}" && !this.lastLoadListJson) return
    if (json === this.lastLoadListJson) return

    this.lastLoadListJson = json

    const data = {
      absolute_urls: this.absoluteUrlsValue
    }

    let filtering = false

    for (const key in rawData) {
      if (rawData[key]) {
        filtering = true
        data[key] = rawData[key]
      }
    }

    this.filteringValue = filtering

    this.listLoadingValue = true

    if (this.listAbortController) {
      this.listAbortController.abort()
    }

    this.listAbortController = new AbortController()

    const url = window.Folio.addParamsToUrl(this.apiListUrlValue, data)

    window.Folio.Api.apiGet(url, null, this.listAbortController.signal).then((res) => {
      this.listContentTarget.innerHTML = res.data || ''
      this.listLoadingValue = false
    }).catch((e) => {
      this.listLoadTimeout = window.setTimeout(() => {
        this.loadList()
      }, 1000)
    }).finally(() => {
      delete this.listAbortController
    })
  }

  cancelFilters () {
    this.formTarget.reset()

    for (const input of this.formTarget.querySelectorAll('.f-input--collection-remote-select')) {
      window.Folio.Input.CollectionRemoteSelect.clearValue(input)
    }

    this.markSelectsAndLoadList()
  }

  markSelectsAndLoadList () {
    for (const input of this.formTarget.querySelectorAll('.f-input--collection-remote-select')) {
      const wrap = input.closest('.f-c-links-modal-url-picker__list-filter')
      wrap.classList.toggle('f-c-links-modal-url-picker__list-filter--active', !!input.value)
    }

    this.loadList()
  }

  resetAdditionalFilter (e) {
    e.preventDefault()
    const input = e.target.closest('.f-c-links-modal-url-picker__list-filter').querySelector('.f-input--collection-remote-select')

    window.Folio.Input.CollectionRemoteSelect.clearValue(input)
  }
})
